### PR TITLE
Extensions protected urls 4928

### PR DESCRIPTION
**Pull request by fmarier**
Draft PR to fix brave/brave-browser#4928.

@bridiver Do you think an override is the best way to patch this?

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [x] Linux
  - [ ] macOS
  - [ ] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [ ] Linux
  - [ ] macOS
  - [ ] Windows
- [ ] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [ ] Ran `git rebase master` (if needed).
- [ ] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [ ] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protection-Mode
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Test Plan:

The webRequest hack can manually tested using https://github.com/fmarier/webrequest-demo and the contentScript one using https://github.com/fmarier/contentscript-demo.

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

File: patches/extensions-common-permissions-permissions_data.cc.patch [hunk 1]
```diff
@@ -0,0 +1,16 @@
+diff --git a/extensions/common/permissions/permissions_data.cc b/extensions/common/permissions/permissions_data.cc
+index b09be07c0db06ec79044ac2468488ad70fce5118..69ed8bdf0743dcdc8672167f246853546aa3073f 100644
+--- a/extensions/common/permissions/permissions_data.cc
++++ b/extensions/common/permissions/permissions_data.cc
+@@ -521,6 +521,11 @@ PermissionsData::PageAccess PermissionsData::CanRunOnPage(
+   if (IsRestrictedUrl(document_url, error))
+     return PageAccess::kDenied;
+ 
++  if (IsBraveRestrictedUrl(document_url, extension_id_, location_, error)) {
```
Comment (bridiver) [inline patches/extensions-common-permissions-permissions_data.cc.patch]: per dm this patch can be removed by using an ExtensionRegistryObserver to add the withheld patterns in either `OnExtensionLoaded` or `OnExtensionReady`. At that point we can check `extensions::PermissionsData::CanExecuteScriptEverywhere` for the extension and decide whether or not to add it and no patch will be necessary

File: patches/extensions-browser-api-web_request-web_request_permissions.cc.patch [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+ #include "base/strings/stringprintf.h"
++#include "brave/extensions/common/brave_extension_urls.h"
+ #include "content/public/browser/child_process_security_policy.h"
+ #include "content/public/browser/resource_request_info.h"
+ #include "extensions/browser/api/extensions_api_client.h"
+@@ -320,6 +321,8 @@ bool WebRequestPermissions::HideRequest(
+   // for requests from common renderers.
+   if (extension_urls::IsWebstoreUpdateUrl(url) ||
+       extension_urls::IsBlacklistUpdateUrl(url) ||
++      extension_urls::IsBraveProtectedUrl(url::Origin::Create(url),
```
Comment (bridiver) [inline patches/extensions-browser-api-web_request-web_request_permissions.cc.patch]: we should use the `ShouldHideBrowserNetworkRequest` call above for this. Subclass `ChromeExtensionsAPIClient` and use the existing `ChromeExtensionsBrowserClient`  chromium_src override to change `ChromeExtensionsAPIClient` to `BraveExtensionsAPIClient`

File: extensions/common/BUILD.gn [hunk 3]
```diff
@@ -0,0 +1,10 @@
+source_set("common") {
+  sources = [
+    "brave_extension_urls.cc",
+    "brave_extension_urls.h",
+  ]
+
+  deps = [
+    "//base",
```
Comment (bridiver) [inline extensions/common/BUILD.gn]: also just for reference this set of deps is not complete. It should also have `//extensions/common` and `//url` - you won't actually need this anymore, but just for future reference you need a dep for anything in the headers

Comment (bridiver) [reply extensions/common/BUILD.gn]: `//url` - https://github.com/brave/brave-core/pull/2779/files#diff-13ddefeda1421d0383faa29f42096c79R8
`//extensions/common` - https://github.com/brave/brave-core/pull/2779/files#diff-13ddefeda1421d0383faa29f42096c79R7

**Comment by fmarier** [comment]
Superseded by #2946

**Pull request closed without merging**